### PR TITLE
deprecate and disable by default DynamicKubeletConfig feature flag

### PR DESCRIPTION
**Pull request by SergeyKanzhelev**
#### What type of PR is this?

Deprecation messages for DynamicKubeletConfig (KEP: https://github.com/kubernetes/enhancements/issues/281).

/sig node
/kind cleanup
/kind documentation
/kind deprecation

#### Which issue(s) this PR fixes:
Relates #100799

(actual issue will need more clean up work)

#### Special notes for your reviewer:
I ***think*** I understood the deprecation rules correctly: https://kubernetes.io/docs/reference/using-api/deprecation-policy/#deprecation and I can disable feature gate on deprecation announcement.

#### Does this PR introduce a user-facing change?
```release-note
The feature Dynamic Kubelet Configuration is deprecated and kubelet will report warning when the flag `--dynamic-config-dir` is used. Feature gate `DynamicKubeletConfig` is disabled out of the box and needs to be explicitly enabled.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

- KEP: https://github.com/kubernetes/enhancements/issues/281

### CODE REVIEW

File: cluster/gce/util.sh [hunk 1]
```diff
@@ -762,9 +762,6 @@ function construct-linux-kubelet-flags {
   flags+=" --experimental-check-node-capabilities-before-mount=true"
   # Keep in sync with the mkdir command in configure-helper.sh (until the TODO is resolved)
   flags+=" --cert-dir=/var/lib/kubelet/pki/"
-  # Configure the directory that the Kubelet should use to store dynamic config checkpoints
-  flags+=" --dynamic-config-dir=/var/lib/kubelet/dynamic-config"
```
Comment (mtaufen) [inline cluster/gce/util.sh:766]: Double check that this doesn't break the node e2e tests.

File: pkg/features/kube_features.go [hunk 2]
```diff
@@ -739,7 +740,7 @@ func init() {
 // available throughout Kubernetes binaries.
 var defaultKubernetesFeatureGates = map[featuregate.Feature]featuregate.FeatureSpec{
 	AppArmor:             {Default: true, PreRelease: featuregate.Beta},
-	DynamicKubeletConfig: {Default: true, PreRelease: featuregate.Beta},
+	DynamicKubeletConfig: {Default: false, PreRelease: featuregate.Deprecated}, // feature gate is deprecated in 1.22, remove no early than 1.23
```
Comment (mtaufen) [inline pkg/features/kube_features.go:743]: This is a breaking change. We should provide the deprecation warning for at least one release before we disable the feature gate. The deprecation policy has gotten even more complex since I last read it 😁 , but I think this is the part we want (where setting to `false` is the last thing we do, but we can set `featuregate.Deprecated` now):

> In some cases, removing pre-GA features requires considerable time. Feature gates can remain operational until their associated feature is fully removed, at which point the feature gate itself can be deprecated.

> Versioning for feature gates is different from the previously discussed components, therefore the rules for deprecation are as follows:

> Rule #8: Feature gates must be deprecated when the corresponding feature they control transitions a lifecycle stage as follows. Feature gates must function for no less than:

> Beta feature to EOL: 3 months or 1 release (whichever is longer)

> Rule #9: Deprecated feature gates must respond with a warning when used. When a feature gate is deprecated it must be documented in both in the release notes and the corresponding CLI help. Both warnings and documentation must indicate whether a feature gate is non-operational.

We should also mark the `Node.Spec.ConfigSource` field deprecated: https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/api/core/v1/types.go#L4610

FWIW, we might want to consider a 2 or 3 release timeline since that field is in core/v1 (it's technically beta, and the doc comment in core/v1 still mentions a feature gate is required, but might still be best to play it safe). @liggitt WDYT?

File: cluster/gce/util.sh (hunk 1 above)
Comment (mtaufen) [reply cluster/gce/util.sh:766]: Well, I guess they passed, so ok then :D.

Given the whole `cluster/` directory is deprecated, I'm not sure we offer any guarantees for these scripts, but I do wonder if anyone might be depending on this who would be broken, and who might appreciate some forewarning.

Comment (SergeyKanzhelev) [reply cluster/gce/util.sh:766]: e2e tests will also set directory when the feature flag is set. Which is safer.

I understand the risk here. But since it only suppose to be used in e2e tests, I'd suggest to enforce it now. Reverting will be straightforward

File: pkg/features/kube_features.go (hunk 2 above)
Comment (SergeyKanzhelev) [reply pkg/features/kube_features.go:743]: I read these rules as the feature gate should remain operational, but it says nothing about the defaults. So I kept it operational and for increased visibility into the change - disabled it. I understand that on the cluster upgrade, if wasn't re-enabled explicitly, it will stop working. But this is a beta feature and my assumption was that "Features can be removed at any point in the life cycle prior to GA". I'm happy to revert the change of default, I see arguments both ways.

I will add deprecation message to the fields, good catch. I was also thinking if `kubectl` should warn when these fields are used, but didn't find any precedence doing so. WDYT?

**Comment by SergeyKanzhelev** [comment]
/retest

failure looks unrelated

File: pkg/features/kube_features.go (hunk 2 above)
Comment (liggitt) [reply pkg/features/kube_features.go:743]: > We should provide the deprecation warning for at least one release before we disable the feature gate.

Someone who still needs to use it for the duration of the deprecation period will turn it on explicitly in 1.22, right? Someone who doesn't will not start accidentally using it in 1.22 without realizing it is beta. I would tend to change the default to false early to increase visibility.

Comment (mtaufen) [reply pkg/features/kube_features.go:743]: > I would tend to change the default to false early to increase visibility.

I suppose that's fair, since it's most of our warning mechanisms (logs, help text, emails to kubernetes-dev) go ignored anyway...

File: cluster/gce/util.sh (hunk 1 above)
Comment (mtaufen) [reply cluster/gce/util.sh:766]: Ok, SGTM

File: staging/src/k8s.io/api/core/v1/types.go [hunk 3]
```diff
@@ -4616,6 +4618,7 @@ type NodeSpec struct {
 }
 
 // NodeConfigSource specifies a source of node configuration. Exactly one subfield (excluding metadata) must be non-nil.
+// This API is deprecated since 1.11
```
Comment (mtaufen) [inline staging/src/k8s.io/api/core/v1/types.go:4621]: This should be 1.22 also, right?

File: staging/src/k8s.io/api/core/v1/types.go [hunk 4]
```diff
@@ -4604,8 +4604,10 @@ type NodeSpec struct {
 	// If specified, the node's taints.
 	// +optional
 	Taints []Taint `json:"taints,omitempty" protobuf:"bytes,5,opt,name=taints"`
-	// If specified, the source to get node configuration from
+
+	// Deprecated. If specified, the source to get node configuration from
 	// The DynamicKubeletConfig feature gate must be enabled for the Kubelet to use this field
+	// This field is deprecated as of 1.22: https://git.k8s.io/enhancements/keps/sig-node/281-dynamic-kubelet-configuration
```
Comment (liggitt) [inline staging/src/k8s.io/api/core/v1/types.go:4610]: add a check in nodeStrategy#WarningsOnCreate and nodeStrategy#WarningsOnUpdate if this field is non-nil to return a warning:

```
spec.configSource: deprecated in v1.22+, support removal is planned in v1.X
```

File: cmd/kubelet/app/options/options.go [hunk 5]
```diff
@@ -334,7 +334,8 @@ func (f *KubeletFlags) AddFlags(mainfs *pflag.FlagSet) {
...
 
-	fs.Var(&f.DynamicConfigDir, "dynamic-config-dir", "The Kubelet will use this directory for checkpointing downloaded configurations and tracking configuration health. The Kubelet will create this directory if it does not already exist. The path may be absolute or relative; relative paths start at the Kubelet's current working directory. Providing this flag enables dynamic Kubelet configuration. The DynamicKubeletConfig feature gate must be enabled to pass this flag; this gate currently defaults to true because the feature is beta.")
+	fs.Var(&f.DynamicConfigDir, "dynamic-config-dir", "The Kubelet will use this directory for checkpointing downloaded configurations and tracking configuration health. The Kubelet will create this directory if it does not already exist. The path may be absolute or relative; relative paths start at the Kubelet's current working directory. Providing this flag enables dynamic Kubelet configuration. The DynamicKubeletConfig feature gate must be enabled to pass this flag.")
+	fs.MarkDeprecated("dynamic-config-dir", "Feature is deprecated and will not move to GA. Please use alternative ways to update kubelet configuration.")
```
Comment (liggitt) [inline cmd/kubelet/app/options/options.go:338]: be explicit in the message about the starting and targeted removal release

**Comment by liggitt** [comment]
one comment about adding a field deprecation warning, lgtm otherwise

File: cmd/kubelet/app/options/options.go (hunk 5 above)
Comment (SergeyKanzhelev) [reply cmd/kubelet/app/options/options.go]: updated the message

File: staging/src/k8s.io/api/core/v1/types.go (hunk 4 above)
Comment (SergeyKanzhelev) [reply staging/src/k8s.io/api/core/v1/types.go:4609]: added, thank you for suggestion

File: staging/src/k8s.io/api/core/v1/types.go (hunk 3 above)
Comment (SergeyKanzhelev) [reply staging/src/k8s.io/api/core/v1/types.go]: fixed

**Comment by dchen1107** [comment]
/lgtm
/approve

We called out to the community for deprecating this feature due to lack of the user inputs, support and maintenance in the past. Now we finally made the decision. Thanks for the effort.

**Comment by SergeyKanzhelev** [comment]
@kubernetes/api-approvers can you please approve this PR?

File: staging/src/k8s.io/api/core/v1/types.go [hunk 6]
```diff
@@ -4603,8 +4603,10 @@ type NodeSpec struct {
 	// If specified, the node's taints.
 	// +optional
 	Taints []Taint `json:"taints,omitempty" protobuf:"bytes,5,opt,name=taints"`
-	// If specified, the source to get node configuration from
+
+	// Deprecated. If specified, the source to get node configuration from
```
Comment (lavalamp) [inline staging/src/k8s.io/api/core/v1/types.go:4607]:
```suggestion
	// Deprecated. If specified, the source of the node's configuration.
```

Otherwise it is a very confusing run-on sentence (documentation doesn't necessarily preserve line breaks).

Comment (lavalamp) [reply staging/src/k8s.io/api/core/v1/types.go:4607]: The next line also needs a period.

File: staging/src/k8s.io/kubelet/config/v1beta1/types.go [hunk 7]
```diff
@@ -90,21 +90,21 @@ type KubeletConfiguration struct {
 
 	// enableServer enables Kubelet's secured server.
 	// Note: Kubelet's insecure port is controlled by the readOnlyPort option.
-	// Dynamic Kubelet Config (beta): If dynamically updating this field, consider that
+	// Dynamic Kubelet Config (deprecated): If dynamically updating this field, consider that
```
Comment (lavalamp) [inline staging/src/k8s.io/kubelet/config/v1beta1/types.go:93]: I'm not sure I see the reason for editing something in a beta api like this?

Comment (SergeyKanzhelev) [reply staging/src/k8s.io/kubelet/config/v1beta1/types.go:93]: the `(beta)` here indicates the status of the `DynamicKubeletConfig` feature. I suggest to highlight the fact that the feature will be gone soon. Alternatively, I can change `(deprecated)` to something like `to be removed`?

File: staging/src/k8s.io/api/core/v1/types.go (hunk 6 above)
Comment (SergeyKanzhelev) [reply staging/src/k8s.io/api/core/v1/types.go]: I'll pay attention next time. Thank you!

File: staging/src/k8s.io/kubelet/config/v1beta1/types.go (hunk 7 above)
Comment (lavalamp) [reply staging/src/k8s.io/kubelet/config/v1beta1/types.go:93]: Presumably in v1beta1 it was beta, though? I don't know if I object, this just seems like modifying the past...

Comment (SergeyKanzhelev) [reply staging/src/k8s.io/kubelet/config/v1beta1/types.go:93]: I understand. It is a valid point. However, the problem is that when feature will be removed, this API behavior will also change retroactively. This is why I think it makes sense to update the description now.

**Comment by SergeyKanzhelev** [comment]
looks unrelated, let me try again

/retest

File: staging/src/k8s.io/kubelet/config/v1beta1/types.go (hunk 7 above)
Comment (lavalamp) [reply staging/src/k8s.io/kubelet/config/v1beta1/types.go:93]: OK

**Comment by lavalamp** [comment]
/approve

for API bits.

File: staging/src/k8s.io/kubelet/config/v1beta1/types.go (hunk 7 above)
Comment (ehashman) [reply staging/src/k8s.io/kubelet/config/v1beta1/types.go:93]: I found the `Dynamic Kubelet Config (deprecated)` comments here confusing, because for a moment I thought it implied the `KubeletConfig` was being deprecated.

@SergeyKanzhelev I would suggest that we _remove_ all mention of DynamicKubeletConfig from the KubeletConfig comments here instead. Since it's being deprecated, IMO this advice is no longer relevant, and since KubeletConfig is a beta API I think we can change this.

**Review by ehashman — CHANGES_REQUESTED** [review]

File: staging/src/k8s.io/kubelet/config/v1beta1/types.go (hunk 7 above)
Comment (lavalamp) [reply staging/src/k8s.io/kubelet/config/v1beta1/types.go:93]: I think it's good to document the current state even if that is confusing. I agree it could be worded better, e.g. "If DynamicKubeletConfig (deprecated; default off) is on, then ..."

I don't feel too strongly. The pre-existing sentences are not super clear, either.

Comment (SergeyKanzhelev) [reply staging/src/k8s.io/kubelet/config/v1beta1/types.go:93]: I can send the proposed clarified text as a separate PR.

**Comment by SergeyKanzhelev** [comment]
@ehashman would you be OK to lgtm this PR?

**Review by ehashman — COMMENTED** [review]
/lgtm

acking clarification in separate PR
